### PR TITLE
LibWeb: Support Object-position CSS property

### PR DESCRIPTION
**Pull request by TobyAsE**
This PR add support for the object-position property.
As I recently built the object-fit property, this seemed like a logical next step.
With object-position you can tell your image how you want to paint it, very useful in combination with the object-fit!

I introduced the object-position property and a "Position Identifier", which can hold the different identifiers that are valid for the property. The rendering is then just a sea of pain with offsets and pixel-math.
This implementation looks the same in Ladybird and Firefox :^)

image
image

### CODE REVIEW

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
I think this wants to work a bit differently:
- Allow `"valid-types": [ "position" ]` in the JSON. (That means renaming the `position` enum we have to something else unfortunately.) It should parse as a `PositionStyleValue`, which we already have. `<position>` is a general type in VALUES-4 so we should allow it as such.
- Then, we can just say "object-fit takes a position" and won't need to decode things in `StyleProperties::object_position()`.

Otherwise, `StyleProperties::object_position()` is basically just doing parsing but not in the parser. :^)

Also, am I right in thinking that ObjectPosition could just hold LengthPercentages? `left` and `top` should be equivalent to `0%`, `right` and `bottom` to `100%`, and `center` to `50%` if I'm understanding correctly. That would make the logic simpler in a few places.

File: Userland/Libraries/LibWeb/CSS/Properties.json [hunk 1]
```diff
@@ -1624,6 +1624,18 @@
       "object-fit"
     ]
   },
+  "object-position": {
```
[the remark this answers is not in the document]
Comment (TobyAsE) [reply Userland/Libraries/LibWeb/CSS/Properties.json:1627]: That's embarrassing :^)

File: Userland/Libraries/LibWeb/Painting/ImagePaintable.cpp [hunk 2]
```diff
@@ -115,17 +114,76 @@ void ImagePaintable::paint(PaintContext& context, PaintPhase phase) const
...
-            bitmap_intersect.set_x((bitmap_rect.width() - bitmap_intersect.width()) / 2);
-            bitmap_intersect.set_y((bitmap_rect.height() - bitmap_intersect.height()) / 2);
-
-            auto offset_x = (image_int_rect.width() - bitmap_rect.width() * scale_x) / 2;
-            auto offset_y = (image_int_rect.height() - bitmap_rect.height() * scale_y) / 2;
+            auto scaled_bitmap_width = bitmap_rect.width() * scale_x;
+            auto scaled_bitmap_height = bitmap_rect.height() * scale_y;
+
+            auto object_position = dom_element.computed_css_values()->object_position();
+            auto offset_x = 0;
```
[the remark this answers is not in the document]
Comment (TobyAsE) [reply Userland/Libraries/LibWeb/Painting/ImagePaintable.cpp]: They get casted to int further down anyway because of the intrect that is the draw_rect

**Comment by TobyAsE** [comment]
@AtkinsSJ 
Thank you, you of course had the correct idea!
I changed the PR to incorporate your suggestions.

File: Userland/Libraries/LibWeb/CSS/Properties.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (AtkinsSJ) [reply Userland/Libraries/LibWeb/CSS/Properties.json:1627]: Commit title still says it! :laughing:

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
Looks good now except the commit title. :wink: 

We shouldn't need a position enum, but I see why we do at the moment because of how `background-position` parsing works. I'll make a note to untangle that a bit later. :sweat_smile:

**Comment by TobyAsE** [comment]
The title, of course!
All fixed now :^)
Has your yakstack already reached the moon?

**Comment by AtkinsSJ** [comment]
> Has your yakstack already reached the moon?

I can neither confirm or deny the existence of the Yak Space Program.

**Pull request merged**